examples/AdvancedOperations/AddSmartDisplayAd.php

```diff
@@ -0,0 +1,436 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Ads\GoogleAds\Examples\AdvancedOperations;
+
+require __DIR__ . '/../../vendor/autoload.php';
+
+use GetOpt\GetOpt;
+use Google\Ads\GoogleAds\Examples\Utils\ArgumentNames;
+use Google\Ads\GoogleAds\Examples\Utils\ArgumentParser;
+use Google\Ads\GoogleAds\Lib\OAuth2TokenBuilder;
+use Google\Ads\GoogleAds\Lib\V2\GoogleAdsClient;
+use Google\Ads\GoogleAds\Lib\V2\GoogleAdsClientBuilder;
+use Google\Ads\GoogleAds\Lib\V2\GoogleAdsException;
+use Google\Ads\GoogleAds\V2\Common\AdImageAsset;
+use Google\Ads\GoogleAds\V2\Common\AdTextAsset;
+use Google\Ads\GoogleAds\V2\Common\ImageAsset;
+use Google\Ads\GoogleAds\V2\Common\ResponsiveDisplayAdInfo;
+use Google\Ads\GoogleAds\V2\Common\TargetCpa;
+use Google\Ads\GoogleAds\V2\Enums\AdGroupAdStatusEnum\AdGroupAdStatus;
+use Google\Ads\GoogleAds\V2\Enums\AdGroupStatusEnum\AdGroupStatus;
+use Google\Ads\GoogleAds\V2\Enums\AdvertisingChannelSubTypeEnum\AdvertisingChannelSubType;
+use Google\Ads\GoogleAds\V2\Enums\AdvertisingChannelTypeEnum\AdvertisingChannelType;
+use Google\Ads\GoogleAds\V2\Enums\AssetTypeEnum\AssetType;
+use Google\Ads\GoogleAds\V2\Enums\BudgetDeliveryMethodEnum\BudgetDeliveryMethod;
+use Google\Ads\GoogleAds\V2\Errors\GoogleAdsError;
+use Google\Ads\GoogleAds\V2\Resources\Ad;
+use Google\Ads\GoogleAds\V2\Resources\AdGroup;
+use Google\Ads\GoogleAds\V2\Resources\AdGroupAd;
+use Google\Ads\GoogleAds\V2\Resources\Asset;
+use Google\Ads\GoogleAds\V2\Resources\Campaign;
+use Google\Ads\GoogleAds\V2\Resources\CampaignBudget;
+use Google\Ads\GoogleAds\V2\Services\AdGroupAdOperation;
+use Google\Ads\GoogleAds\V2\Services\AdGroupOperation;
+use Google\Ads\GoogleAds\V2\Services\AssetOperation;
+use Google\Ads\GoogleAds\V2\Services\CampaignBudgetOperation;
+use Google\Ads\GoogleAds\V2\Services\CampaignOperation;
+use Google\Ads\GoogleAds\V2\Services\MutateAdGroupAdsResponse;
+use Google\Ads\GoogleAds\V2\Services\MutateAdGroupsResponse;
+use Google\Ads\GoogleAds\V2\Services\MutateCampaignBudgetsResponse;
+use Google\Ads\GoogleAds\V2\Services\MutateCampaignsResponse;
+use Google\ApiCore\ApiException;
+use Google\Protobuf\BytesValue;
+use Google\Protobuf\Int64Value;
+use Google\Protobuf\StringValue;
+
+/**
+ * This example adds a Smart Display campaign, an ad group and a responsive display ad.
+ * More information about Smart Display campaigns can be found at
+ * https://support.google.com/google-ads/answer/7020281.
+ *
+ * IMPORTANT: The AssetService requires you to reuse what you've uploaded previously. Therefore,
```

Comment:
> This might become a blocking issue when there will be more than one code example using assets that loads the same images

Why? We can share assets and images across ads and across code examples. Could you elaborate?

Reply:
Note that 2) is not feasible especially for images, AssetService translates the URL we've uploaded into something else (I guess it normalizes all the URLs), so you cannot just use the URL you use to query.

Reply:
> > This might become a blocking issue when there will be more than one code example using assets that loads the same images
> 
> Why? We can share assets and images across ads and across code examples. Could you elaborate?

If I understand well your comment, it would be impossible to create multiple assets using the same URL (image bytes). So if I run the AddSmartDisplayAd and then the AddGmailAd code examples both with default image URLs, the AddGmailAd code example will fail because the same image URLs are being used.

I am suggesting to somehow make it smoother for the developers either by adding comments or to workaround this constraint programmatically in all the code examples that share the same image URLs.

Reply:
> If I understand well your comment, it would be impossible to create multiple assets using the same URL (image bytes). So if I run the AddSmartDisplayAd and then the AddGmailAd code examples both with default image URLs, the AddGmailAd code example will fail because the same image URLs have been used before.

I believe that's a different system.
If AddGmailAd uses assets, the user should specify the existing asset resource names.
If not (like using a MediaService's created image), they're not related and this shouldn't block them from uploading the same image.

Reply:
Ok thanks for clarifying.